File: javascript/simple-cipher/simple-cipher.js
const SMALL_LETTER_SHIFT = 97;
const NUM_LETTERS = 26;
const RANDOM_KEY_LENGTH = 100;

const DIRECTION = {
  ENCODE: 1,
  DECODE: -1
}


export class Cipher {
  constructor(key = this.#generate_key(RANDOM_KEY_LENGTH)) {
    if (!/^[a-z]+$/.test(key)) throw new Error('Bad key')
    this._key = key;
    this._key_length = this._key.length;
  }

  encode(text) {
    return this.#to_code(text, DIRECTION.ENCODE);
  }

  decode(code) {
    return this.#to_code(code, DIRECTION.DECODE);
  }

  get key() {
    return this._key;
  }

  // @param {number} direction - 1 for encode and -1 for decode
  #to_code(code, direction){
    return [...code].map((val, index) =>
      Cipher.index_to_letter(
        Cipher.shift_char_code(val, this._key[index % this._key_length], direction)
      )).join('');
  }

  #generate_key(length) {
    return Array.from({length}, Cipher.random_letter).join('');
  }

  static random_letter = () => Cipher.index_to_letter(0|Math.random() * NUM_LETTERS);

  static letter_to_index = (letter) => letter.charCodeAt(0) - SMALL_LETTER_SHIFT;

  static index_to_letter = (index) => 
    String.fromCharCode(index + SMALL_LETTER_SHIFT);

  static shift_char_code = (char_code, shift_code, direction) =>
    (Cipher.letter_to_index(char_code) + direction * Cipher.letter_to_index(shift_code) + 
      NUM_LETTERS) % NUM_LETTERS
}

File: python/simple-cipher/simple_cipher.py
import random
from itertools import cycle

DEFAULT_KEY_LENGTH = 100
CHAR_A = 97
ALFA_LENGTH = 26


class Cipher:
    def __init__(self, key=None):
        self.key = key
        if not self.key:
            self.key = ''.join([chr(random.randrange(97, 123))
                                for _ in range(0, DEFAULT_KEY_LENGTH)])
        self.key_length = len(self.key)

    def encode(self, text):
        return self.__code(text, 1)

    def decode(self, text):
        return self.__code(text, -1)

    def __code(self, text, side=1):
        return ''.join([chr((ord(char) + side * (ord(key)
                        - CHAR_A) - CHAR_A) % ALFA_LENGTH + CHAR_A)
                        for char, key in zip(text, cycle(self.key))])
